Implement "tiered quantities" for sponsorship benefits and sponsor benefits

ewdurbin opened the pull request:
@berinhard this is my very initial bit at modeling "tiered quantities" for a given `SponsorshipBenefit`.

I'm not _positive_ if I'm using the polymorphic library just right...

TODO:
  - [x] sanity check with Bernardo
  - [x] implement the feature on the resulting `SponsorBenefit`
  - [x] style the benefits form
  - [x] render quantities along side benefits
  - [x] tests

ewdurbin committed:
begin implementing "tiered quantities" for sponsorship benefits and sponsor benefits

ewdurbin committed:
improve display of tiered quantity benefits on application form

ewdurbin committed:
implement "filtering" for TieredBenefit configuration

no need to apply all "levels" to a given benefit

ewdurbin commented on sponsors/models.py:
```diff
@@ -969,6 +973,18 @@ class Meta(BaseTieredQuantity.Meta, BenefitFeatureConfiguration.Meta):
+        return TieredQuantity
+
+    def get_benefit_feature(self, **kwargs):
+        if kwargs["sponsor_benefit"].sponsorship.level_name == self.package.name:
```
thoughts on how to handle the fact that we don't keep a pointer to the source `SponsorshipPackage` around, @berinhard? This seems _OK_ since the package will almost certainly exist with the same name when the application is submitted... but still don't feel great about making it a string comparison.

berinhard replied:
I felt the same after coding the `list_sponsors` template tag. I coded a [similar package name checking](https://github.com/python/pythondotorg/blob/main/sponsors/templatetags/sponsors.py#L38) and I wanted to talk about a refactoring as well.

I'm not sure why we made the decision of not pointing to the `SponsorshipPackage` instance initially but this condition exposes that we need to have it. I can work on a data migration and refactoring to add this to the `Sponsorship` model.

ewdurbin replied:
This approach makes the most sense to me 👍🏼

ewdurbin committed:
display appropriate quantities for tiered quantity benefits throughout

berinhard commented on sponsors/models.py:
```diff
@@ -904,6 +940,9 @@ def benefit_feature_class(self):
     def get_benefit_feature(self, **kwargs):
+        return self.build_benefit_feature(**kwargs)
+
+    def build_benefit_feature(self, **kwargs):
```
This can be confusing for newcomers to this model because this can look like as an unnecessary indirection. I'd suggest to go on with a more Django CBV-like strategy to enable full customization on how to build a benefit feature. We could add a new method such as `get_benefit_feature_kwargs`.

This new method would build the default `kwargs` dict and return it **OR** return an empty dict if no benefit should be created. That way, we could still have the generic logic under the `get_benefit_feature` method as well.

berinhard reviewed (approved):
The code looks good to me, although I left some refactoring considerations. I don't think any of them are deal breakers, but they could be nice to have. 

Let me know if you want any help to code the tests.

berinhard commented on sponsors/models.py:
```diff
@@ -222,9 +229,20 @@ def __str__(self):
+        name = self.name
+        for feature in self.features_config.all():
+            if isinstance(feature, TieredQuantityConfiguration) and feature.package == package:
+                name = feature.display_modifier(name)
```
I think that maybe we can refactor this code as well. We could refactor the `BenefitFeatureConfig.display_modifier` to accept  `**kwargs` alongside with the `name` parameter. That way we could move any type of conditional checking to the specific configuration instance. I think we could end up with something like:


```python
def name_for_display(self, package=None):
    name = self.name
    for feature in self.features_config.all():
        name = feature.display_modifier(name, package=package)
    return name
```

And, then, for the `TieredQuantityConfiguration.display_modifier`

```python
def display_modifier(name, **kwargs):
    package = kwargs.get("package")
    if package == self.package:
        return f"{name} ({self.quantity})"
    else:
       return name
```

This will be easier for us to unit test as well.

berinhard committed:
Refactor on benefit configuration + unit tests

berinhard committed:
Refactor and unit test display modifier for benefit configuration

berinhard committed:
Unit test display modifiers for sponsor benefit

berinhard committed:
Unit test template tags

berinhard committed:
Use cached property to avoid repeated queries on form

This change reduced the numbers of queries executed by the
sponsorship application html template from 366 to 72. In my local env
the total query time reduced from 65.96ms to 27.50ms.

ewdurbin committed:
reorder migration
